### PR TITLE
#594 - Configurable public site languages

### PR DESCRIPTION
**Pull request by blms**
## What this PR does

- Per #594:
  - Uses the `PUBLIC_SITE_LANGUAGES` setting to determine enabled languages for anonymous users
  - Hides disabled languages from the UI
  - Redirects to default language when attempting to reach a disabled one

### CODE REVIEW

File: geniza/common/middleware.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    def process_request(self, request: HttpRequest):
+        public_site_languages = getattr(settings, "PUBLIC_SITE_LANGUAGES", [])
+        if public_site_languages and not request.user.is_authenticated:
+            language_from_path = translation.get_language_from_path(request.path_info)
+            if language_from_path and language_from_path not in public_site_languages:
+                # If the path language isn't in the public site languages list,
+                # update cookies and the request object to use the default language
+                request.COOKIES[settings.LANGUAGE_COOKIE_NAME] = settings.LANGUAGE_CODE
+                translation.activate(settings.LANGUAGE_CODE)
+                request.LANGUAGE_CODE = translation.get_language()
```
Comment (rlskoeser) [inline geniza/common/middleware.py:23]: surprised we have to do the explicit change, is there no way to use the `set_language` redirect view? https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#the-set-language-redirect-view

**Review by rlskoeser — CHANGES_REQUESTED** [review]
Looks really good and works well! Just a couple of small fixes.

Why do we need both `process_request` and `process_response`? Is there no way to do what's needed in a single place?

File: geniza/context_processors.py [hunk 2]
```diff
@@ -10,6 +10,7 @@ def template_globals(request):
     context_extras = {
         "SHOW_TEST_WARNING": getattr(settings, "SHOW_TEST_WARNING", False),
         "FONT_URL_PREFIX": getattr(settings, "FONT_URL_PREFIX", ""),
+        "PUBLIC_SITE_LANGUAGES": getattr(settings, "PUBLIC_SITE_LANGUAGES", []),
```
[the remark this answers is not in the document]
Comment (blms) [reply geniza/context_processors.py]: Done in 042a0ff :+1:

File: geniza/templates/snippets/language_switcher.html [hunk 3]
```diff
@@ -8,14 +8,17 @@
...
                 {% for language in languages %}
-                    <option value="{{ language.code }}"{% if language.code == LANGUAGE_CODE %} selected{% endif %}>
-                        {% language language.code %} {# display label for language in that language! #}
-                            {# Translators: label for language choices in navigation #}
-                            {% blocktranslate with language_name=language.name_local language_code=language.code trimmed %}
-                                read this page in {{ language_name }} ({{ language_code }})
-                            {% endblocktranslate %}
-                        {% endlanguage %}
-                    </option>
+                    {# For anonymous users, only display languages in PUBLIC_SITE_LANGUAGES (if defined) #}
```
[the remark this answers is not in the document]
Comment (blms) [reply geniza/templates/snippets/language_switcher.html]: Done in 001b02c 👍

File: geniza/common/middleware.py (hunk 1 above)
Comment (blms) [reply geniza/common/middleware.py]: Good call, I forgot about `set_language`! Done in d9a7304 👍

File: geniza/common/middleware.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+
+class PublicLocaleMiddleware(MiddlewareMixin):
+    """Middleware to redirect anonymous users attempting to access locales that are not in
+    the list of public site languages."""
+
+    response_redirect_class = HttpResponseRedirect
+
+    def process_request(self, request: HttpRequest):
+        public_site_languages = getattr(settings, "PUBLIC_SITE_LANGUAGES", [])
```
[the remark this answers is not in the document]
Comment (blms) [reply geniza/common/middleware.py]: No longer needed in multiple methods as of d9a7304 👍

**Comment by blms** [comment]
In d9a7304 I ended up using the new middleware style for this, so I'm using `__call__` instead of processing the request and response separately. It seemed to simplify things quite a bit. Let me know what you think!

File: geniza/common/middleware.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                default_language = settings.LANGUAGE_CODE
+                language_path = request.path_info.replace(
+                    language_from_path, default_language
+                )
+                # set request method to POST and pass default language, new path to set_language
+                request.method = "POST"
+                request_data = {}
+                request_data["language"] = default_language
+                request_data["next"] = language_path
+                request.POST = request_data
```
Comment (rlskoeser) [inline geniza/common/middleware.py:29]: can't we do this in one line?
```suggestion
                request.POST = {"language": default_language, "next": language_path}
```

File: geniza/common/middleware.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                return set_language(request)
+
+        # otherwise, continue on with middleware chain
+        response = self.get_response(request)
+        return response
```
Comment (rlskoeser) [inline geniza/common/middleware.py:34]: nice! I'm glad it worked to use `set_language`

**Review by rlskoeser — APPROVED** [review]
Looks great! Nice work refactoring the middleware. Just one small suggestion to consolidate multiple lines of code.

File: geniza/common/middleware.py (hunk 5 above)
Comment (blms) [reply geniza/common/middleware.py:29]: Of course! Great catch 😆 leftovers from when I was trying to create a `QueryDict`…

Comment (rlskoeser) [reply geniza/common/middleware.py:29]: oh, of course! yeah, I remember `QueryDict` being kind of a pain

**Pull request merged**